packages/core-app-api/src/apis/implementations/FetchApi/IdentityAwareFetchMiddleware.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FetchFunction, FetchMiddleware } from './types';
+
+const DEFAULT_HEADER_NAME = 'backstage-token';
+
+/**
+ * A fetch middleware, whose signed-in state can change throughout its lifetime
+ * and injects a Backstage token header accordingly.
+ *
+ * @public
+ */
+export class IdentityAwareFetchMiddleware implements FetchMiddleware {
+  private headerName: string;
+  private tokenFunction: (() => Promise<string | undefined>) | undefined;
+
+  constructor() {
+    this.headerName = DEFAULT_HEADER_NAME;
+    this.tokenFunction = undefined;
+  }
+
+  /**
+   * {@inheritdoc FetchMiddleware.apply}
+   */
+  apply(next: FetchFunction): FetchFunction {
+    return async (input, init) => {
+      // Making sure to read this field from the surrounding class dynamically
+      // on each fetch call, since it might change as the user logs in and out.
```

Comment:
I don't think we need to handle sign-out really. The current contract is to reload the page after sign-out, which I think is going to be difficult to move away from tbh https://github.com/backstage/backstage/blob/040241b061592f0e5920c68da73da7db0cd5f4f0/packages/core-app-api/src/app/AppIdentity.ts#L65

Reply:
Yep, handled this by just passing in the token resolver function directly at construction time instead. This works out better now that we have `AppIdentityProxy`.